lib/IRGen/GenEnum.cpp

```diff
@@ -5845,13 +5845,24 @@ EnumImplStrategy::get(TypeConverter &TC, SILType type, EnumDecl *theEnum) {
     auto origArgLoweredTy = TC.IGM.getLoweredType(origArgType);
     auto *origArgTI = &TC.getCompleteTypeInfo(origArgLoweredTy.getASTType());
 
-    // If the unsubstituted argument contains a generic parameter type, or
-    // is not fixed-size in all resilience domains that have knowledge of
-    // this enum's layout, we need to constrain our layout optimizations to
-    // what the runtime can reproduce.
-    if (!isResilient &&
-        !origArgTI->isFixedSize(layoutScope))
-      allowFixedLayoutOptimizations = false;
+    // If the layout scope is minimal it means that we could have a public fixed
+    // size type that references a resilient type in the same module. If we
+    // lower the resilient type in the normal lowering mode it will be fixed
+    // size because it is local to the module. This is problematic because the
+    // non-resilient type can be accessed from outside the module.
+    {
+      LoweringModeScope scope(TC.IGM,
+                              layoutScope == ResilienceExpansion::Minimal
+                                  ? TypeConverter::Mode::LocallyResilient
+                                  : TypeConverter::Mode::Normal);
+      auto *argTI = &TC.getCompleteTypeInfo(origArgLoweredTy.getASTType());
+      // If the unsubstituted argument contains a generic parameter type, or
+      // is not fixed-size in all resilience domains that have knowledge of
+      // this enum's layout, we need to constrain our layout optimizations to
+      // what the runtime can reproduce.
+      if (!isResilient && !argTI->isFixedSize(layoutScope))
```

Comment:
I don’t understand why you’re adding the new lowering mode. It sounds like isFixedSize() returns true when it should return false?

Reply:
If you lower a resilient type in its local module it ends up as a fixed size type.

Reply:
Right but isFixedSize(ResilienceExpansion::Minimal) is meant to return false if it was fixed size in other resilience domains. It sounds like it’s not being computed correctly in all cases (nested enums?) and we need to fix the root cause instead of papering over it here.

Reply:
Sorry I mean “if it was *not* fixed size in other resilience domains”.

Reply:
The enum that we query in the test case (``enum ResilientType```) is not nested. I have to debug again but what I remember from yesterday is that if I get the type info for a resilient enum inside it's own module it returns a fixed size enum which always has fixed size.

I will confirm again that this is what happens.

Reply:
That is correct. You should get a FixedTypeInfo. However isFixedSize(ResilienceExpansion::Minimal) returns the TypeInfo’s AlwaysFixedSize flag which should be set to false in this case. It sounds like it is not getting set for enums.

Reply:
If you try the same example with a resilient struct instead of a resilient enum, do we correctly emit the payload size (and skip layout optimizations) for the non-resilient enum?

Reply:
Okay, I will look into that. Thanks.

Reply:
Yeah, only happens with the enum. I will track down where we miss setting the flag.

Reply:
Awesome, thank you!

Reply:
Much simpler now.